rust/lance-linalg/src/kmeans.rs

```diff
@@ -538,12 +539,114 @@ impl KMeans {
     }
 }
 
+/// Return a slice of `data[x,y..y+strip]`.
+#[inline]
+fn get_slice(data: &[f32], x: usize, y: usize, dim: usize, strip: usize) -> &[f32] {
+    &data[x * dim + y..x * dim + y + strip]
+}
+
+/// Fast partition computation for L2 distance.
+fn compute_partitions_l2_f32(centroids: &[f32], data: &[f32], dim: usize) -> Vec<u32> {
+    const STRIPE_SIZE: usize = 128;
+    const TILE_SIZE: usize = 16;
+
+    // 128 * 4bytes * 16 = 8KB for centroid and data respectively, so both of them can
+    // stay in L1 cache.
+    let num_rows = data.len() / dim;
+    let num_centroids = centroids.len() / dim;
+
+    let mut partition_ids = vec![0_u32; num_rows];
+
+    // Read a tile of data, `data[idx..idx+TILE_SIZE]`
+    for idx in (0..num_rows).step_by(TILE_SIZE) {
+        let num_rows_in_tile = min(TILE_SIZE, num_rows - idx);
+
+        // Loop over each strip.
+        // s is the index of value in each vector.
+        let mut min_dists = [f32::MAX; TILE_SIZE];
+        for centroid_start in (0..num_centroids).step_by(TILE_SIZE) {
+            // 4B * 16 * 16 = 1 KB
+            let mut dists = vec![0_f32; TILE_SIZE * TILE_SIZE];
+            let num_centroids_in_tile = min(TILE_SIZE, num_centroids - centroid_start);
+            for s in (0..dim).step_by(STRIPE_SIZE) {
+                // Calculate L2 within each TILE * STRIP
+                let slice_len = min(STRIPE_SIZE, dim - s);
+                for di in 0..num_rows_in_tile {
+                    let data_slice = get_slice(data, idx + di, s, dim, slice_len);
+                    for ci in centroid_start..centroid_start + num_centroids_in_tile {
+                        // Get a slice of `data[di][s..s+STRIP_SIZE]`.
+                        let cent_slice = get_slice(centroids, ci, s, dim, slice_len);
+                        let dist = data_slice.l2(cent_slice);
+                        dists[di * TILE_SIZE + (ci - centroid_start)] += dist;
```

Comment:
It's very likely I'm reading this incorrectly but it looks like you are calculating:

```
sqrt(diff_sq(0) + ... + diff_sq(N))
```

by breaking it into pieces:

```
sqrt(diff_sq(0) + ... + diff_sq(a)) + sqrt(diff_sq(a+1) + ... + diff_sq(N))
```

But doesn't this require `sqrt(a + b) = sqrt(a) + sqrt(b)` which isn't true?